src/mozanalysis/experiments.py

```diff
@@ -1,165 +1,268 @@
+            .agg(
+                F.avg("sum_total_hours").alias("engagement_daily_hours"),
+                F.avg("sum_active_hours").alias("engagement_daily_active_hours"),
+                F.expr(
+                    "SUM(sum_total_uris) "
+                    "/ ({hours_epsilon} + SUM(sum_active_hours))".format(**CONSTANTS)
+                ).alias("engagement_hourly_uris"),
+                F.expr(
+                    "SUM(sum_active_hours) "
+                    "/ ({hours_epsilon} + SUM(sum_total_hours))".format(**CONSTANTS)
+                ).alias("engagement_intensity"),
+            )
+        )
 
-        data = []
-        branches = self.get_split_by_values(dataset)
-
-        for m in self._metrics:
-            # TODO: Use a lib or make this more robust.
-            metric_name = m.name.replace(" ", "_").lower()
-
-            agg_df = (
-                dataset.select(*([self._aggregate_by, self._split_by] + m.columns))
-                .groupBy(self._aggregate_by, self._split_by)
-                .agg(*m.aggregations)
-                .select(
-                    *[
-                        self._aggregate_by,
-                        self._split_by,
-                        m.final_expression.alias(metric_name),
-                    ]
-                )
+        return agg_df
+
+    def retention(self, week=3):
+        # Compute only the retention metrics.
+        # Return pandas dataframe with data and CIs.
+        if not self._df:
+            self._df = self.get_data(metric=CoreMetrics.RETENTION)
+
+        retention_df = (
+            self._df.withColumn(
+                "enrollment_date",
+                F.min("submission_date").over(Window().partitionBy("client_id")),
+            )
+            .withColumn(
+                "sum_total_hours",
+                self._df.subsession_length / CONSTANTS["seconds_per_hour"],
+            )
+            .withColumn(
+                "sum_total_uris",
+                self._df.scalar_parent_browser_engagement_total_uri_count,
+            )
+            .select(
+                "enrollment_date",
+                "sum_total_hours",
+                "sum_total_uris",
+                F.datediff("submission_date", "enrollment_date").alias("nd"),
+                F.floor(F.datediff("submission_date", "enrollment_date") / 7).alias(
+                    "week"
+                ),
+                *self._df.columns
+            )
+            .filter(F.col("nd") >= 0)
+        )
+
+        week3_df = (
+            retention_df.filter("week={week}".format(week=week))
+            .groupBy(*AGG_COLS)
+            .agg(
+                F.expr("CASE WHEN SUM(sum_total_hours) > 0 THEN 1 ELSE 0 END").alias(
+                    "retention"
+                ),
+                F.expr("MAX(CASE WHEN sum_total_uris >= 5 THEN 1 ELSE 0 END)").alias(
+                    "active_retention"
+                ),
             )
+        )
+
+        return week3_df
```

Comment:
I don't think this tells me what I want to know, which is "what fraction of users that were present in week 0 are still present (and/or active) in week _n_"; this gives me a list of client_ids that were present in week 3?

Reply:
I think you may be trying to produce something to bootstrap over? Filtering to the _n_th week means that this isn't a useful thing to bootstrap over (instead of joining a column onto the list of clients that represents whether they were present or active in week _n_, which would be) but also we can use the normal approximation for a binomial confidence interval, and save ourselves some computing: https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Normal_approximation_interval

Reply:
(this difference may be another good reason to segregate the engagement and retention metrics)

Reply:
Thanks for pointing this out. At one point this code was the Python equivalent of sguha's SQL, but with multiple refactorings something must have gotten lost along the way. I'll go back and review this.

Reply:
It appears that my Python is roughly equivalent to the SQL with one exception... at the end the SQL does a JOIN of the retention data with the engagement metrics. The subtlety that was missed was that any rows now in the retention table would JOIN with a zero in the retention columns. That is lacking from my implementation.

A proposed solution is to calculate each week's retention as separate columns, e.g. :`retained_in_week_1`, `retained_in_week_2`, `retained_in_week_3`, and likewise for active retention, so that the equivalent comparison could be made and also satisfy Tim's original comment in this thread.

Thoughts?